llama_index/postprocessor/rankGPT_rerank.py

```diff
@@ -0,0 +1,197 @@
+
+            for idx in rerank_ranks:
+                initial_results.append(
+                    NodeWithScore(node=nodes[idx].node, score=nodes[idx].score)
+                )
+            return initial_results[: self.top_n]
+        else:
+            return nodes[: self.top_n]
+
+    def _get_prefix_prompt(self, query: str, num: int) -> List[ChatMessage]:
+        return [
+            ChatMessage(
+                role="system",
+                content="You are RankGPT, an intelligent assistant that can rank passages based on their relevancy to the query.",
+            ),
+            ChatMessage(
+                role="user",
+                content=f"I will provide you with {num} passages, each indicated by number identifier []. \nRank the passages based on their relevance to query: {query}.",
+            ),
+            ChatMessage(role="assistant", content="Okay, please provide the passages."),
+        ]
+
+    def _get_post_prompt(self, query: str, num: int) -> str:
+        return f"Search Query: {query}. \nRank the {num} passages above based on their relevance to the search query. The passages should be listed in descending order using identifiers. The most relevant passages should be listed first. The output format should be [] > [], e.g., [1] > [2]. Only response the ranking results, do not say any word or explain."
+
+    def create_permutation_instruction(self, item: Dict[str, Any]) -> List[ChatMessage]:
+        query = item["query"]
+        num = len(item["hits"])
+
+        max_length = 300
+        while True:
+            messages = self._get_prefix_prompt(query, num)
+            rank = 0
+            for hit in item["hits"]:
+                rank += 1
+                content = hit["content"]
+                content = content.replace("Title: Content: ", "")
+                content = content.strip()
+                # For Japanese should cut by character: content = content[:int(max_length)]
+                content = " ".join(content.split()[: int(max_length)])
+                messages.append(ChatMessage(role="user", content=f"[{rank}] {content}"))
+                messages.append(
+                    ChatMessage(role="assistant", content=f"Received passage [{rank}].")
+                )
+            messages.append(
+                ChatMessage(role="user", content=self._get_post_prompt(query, num))
+            )
+            context_size_limit = (
+                openai_modelname_to_contextsize(self.model)
+                if "gpt" in self.model
+                else self.context_size
+            )
+            if (
+                self.token_counter.estimate_tokens_in_messages(messages)
+                <= context_size_limit - 200
+            ):
+                break
+            else:
+                max_length -= 1
+        return messages
+
+    def run_llm(self, messages: Sequence[ChatMessage]) -> ChatResponse:
+        if "gpt" in self.model:
+            return OpenAI(
```

Comment:
why we don't just let the user pass in an llm as an optional arg? and default to llm, so remove self.model, self.api_key as init arguments. 

this is more consistent with all of our other interfaces

Reply:
my bad. changed to passing LLM as a parameter.
Also test `Mistral` thru Ollama, which is interestion